File: apps/mo1-ios/src/deeplink.ts
// deeplink — Universal Links (https) primary + `dub://` fallback parsing
// (design §2-3: "Universal Links 本命 + dub:// フォールバック・devhub:// は廃止").
// Resolves a URL to an in-app route the AppShell can navigate to.
export type RouteName = "home" | "event" | "action" | "task" | "inbox" | "profile" | "chat";

export interface Route {
  name: RouteName;
  params: Record<string, string>;
}

const HOME: Route = { name: "home", params: {} };

const UNIVERSAL_HOSTS: ReadonlySet<string> = new Set(["m.developershub.jp", "developershub.jp"]);
const DUB_SCHEME = "dub:";

/** Map a path segment list to a route; unknown paths fall back to home. */
function routeFromSegments(segments: string[]): Route {
  const [head, id] = segments;
  switch (head) {
    case undefined:
    case "":
    case "home":
      return HOME;
    case "events":
      return id ? { name: "event", params: { eventId: id } } : HOME;
    case "actions":
      return id ? { name: "action", params: { actionId: id } } : HOME;
    case "tasks":
      return id ? { name: "task", params: { taskId: id } } : HOME;
    case "inbox":
      return { name: "inbox", params: {} };
    case "chat":
      return id ? { name: "chat", params: { channelId: id } } : { name: "chat", params: {} };
    case "profile":
    case "me":
      return { name: "profile", params: {} };
    default:
      return HOME;
  }
}

/** Parse an https Universal Link or a dub:// deeplink into a Route. */
export function parseDeepLink(raw: string): Route {
  let u: URL;
  try {
    u = new URL(raw);
  } catch {
    return HOME;
  }

  if (u.protocol === DUB_SCHEME) {
    // dub://events/<id>  -> host="events", pathname="/<id>"
    const host = u.hostname;
    const rest = u.pathname.split("/").filter(Boolean);
    return routeFromSegments([host, ...rest]);
  }

  if ((u.protocol === "https:" || u.protocol === "http:") && UNIVERSAL_HOSTS.has(u.hostname)) {
    return routeFromSegments(u.pathname.split("/").filter(Boolean));
  }

  // devhub:// and any foreign origin are ignored (retired scheme / safety).
  return HOME;
}

File: apps/mo1-ios/ios/Sources/Mo1Core/Domain/DeepLink.swift
// DeepLink — Universal Links (https) primary + `dub://` fallback parsing
// (design §2-3: "Universal Links 本命 + dub:// フォールバック・devhub:// は廃止"),
// mirrors deeplink.ts. Resolves a URL to an in-app route.
import Foundation

public enum RouteName: String, Equatable, Sendable {
    case home, event, action, task, inbox, profile, chat
}

public struct Route: Equatable, Sendable {
    public var name: RouteName
    public var params: [String: String]
    public init(name: RouteName, params: [String: String] = [:]) {
        self.name = name; self.params = params
    }
}

private let HOME = Route(name: .home)
private let UNIVERSAL_HOSTS: Set<String> = ["m.developershub.jp", "developershub.jp"]

/// Map a path segment list to a route; unknown paths fall back to home.
private func routeFromSegments(_ segments: [String]) -> Route {
    let head = segments.first
    let id = segments.count > 1 ? segments[1] : nil
    switch head {
    case nil, "", "home":
        return HOME
    case "events":
        return id.map { Route(name: .event, params: ["eventId": $0]) } ?? HOME
    case "actions":
        return id.map { Route(name: .action, params: ["actionId": $0]) } ?? HOME
    case "tasks":
        return id.map { Route(name: .task, params: ["taskId": $0]) } ?? HOME
    case "inbox":
        return Route(name: .inbox)
    case "chat":
        return id.map { Route(name: .chat, params: ["channelId": $0]) } ?? Route(name: .chat)
    case "profile", "me":
        return Route(name: .profile)
    default:
        return HOME
    }
}

/// Parse an https Universal Link or a dub:// deeplink into a Route.
public func parseDeepLink(_ raw: String) -> Route {
    guard let comps = URLComponents(string: raw), let scheme = comps.scheme?.lowercased() else {
        return HOME
    }
    let pathSegments = comps.path.split(separator: "/").map(String.init)

    if scheme == "dub" {
        // dub://events/<id> -> host="events", path="/<id>"
        let host = comps.host ?? ""
        return routeFromSegments([host] + pathSegments)
    }

    if (scheme == "https" || scheme == "http"), let host = comps.host, UNIVERSAL_HOSTS.contains(host) {
        return routeFromSegments(pathSegments)
    }

    // devhub:// and any foreign origin are ignored (retired scheme / safety).
    return HOME
}
